optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolver.java

```diff
@@ -149,16 +149,20 @@ public final void solve(Solution planningProblem) {
             throw new IllegalArgumentException("The planningProblem (" + planningProblem
                     + ") must not be null.");
         }
-        solverScope.setBestSolution(planningProblem);
-        outerSolvingStarted(solverScope);
-        boolean restartSolver = true;
-        while (restartSolver) {
-            solvingStarted(solverScope);
-            runPhases();
-            solvingEnded(solverScope);
-            restartSolver = checkProblemFactChanges();
+        try {
+            solverScope.setBestSolution(planningProblem);
+            outerSolvingStarted(solverScope);
+            boolean restartSolver = true;
+            while (restartSolver) {
+                solvingStarted(solverScope);
+                runPhases();
+                solvingEnded(solverScope);
+                restartSolver = checkProblemFactChanges();
+            }
+            outerSolvingEnded(solverScope);
+        } finally {
+            solving.set(false);
```

Comment:
@snurkabill @dimiii Catching the exception and returning Double.NEGATIVE_INFINITY will effectively eat your exception. That's very bad.
Instead, surround the Solver.solve() method's call with try-catch (in your code) and report it there properly (for example call an BrokenSolverEvent, similar to the BestSolutionChangedEvent (note that that one is called in the solver thread too).

I do agree, that in OptaPlanner's future SolverManager (= a high-level wrapper around the low-level Solver that manages threads too, does load balancing, multitenancy, etc) will need to cover exception handling and notify exceptions correctly.